source/agora/consensus/state/ValidatorSet.d

```diff
@@ -204,24 +172,19 @@ public class ValidatorSet
 
     /***************************************************************************
 
-        Unenroll the enrollment with the given UTXO key from the validator set
-        First we remove any previous expired record for this key and then we set
-        the current active record to expired
+        Remove the enrollment with the given UTXO key from the validator set
 
         Params:
             utxo_key = the UTXO key of the enrollment data to unerno
 
     ***************************************************************************/
 
-    public void unenroll (in Hash enroll_hash) @trusted nothrow
+    public void remove (in Hash enroll_hash) @trusted nothrow
     {
         try
         {
             () @trusted {
-                this.db.execute("DELETE from validator_set WHERE key = ? AND active = ?",
-                    EnrollmentStatus.Expired, enroll_hash);
-                this.db.execute("UPDATE validator_set SET active = ? WHERE key = ?",
-                    EnrollmentStatus.Expired, enroll_hash);
+                this.db.execute("DELETE from validator_set WHERE key = ?", enroll_hash);
             }();
         }
         catch (Exception ex)
```

Comment:
Do we really want this function ? Feels odd to me to have both a primary key that takes into account the enrolled_height and a function to remove entries that is used to essentially keep the count of entries at 1 ?

Reply:
It was only to keep the table row size to what is really needed. We should never need to query enrollments older than the previous cycle. I suppose redundant data cleanup can be done periodically by a task rather than each cycle.

Reply:
Don't we need it for block rewards ?

Reply:
What shall we do with a slashed validator? It was being removed from the validator set which is probably not really what we want. However we do not want to treat it as an active validator after it has been slashed. A slashed flag?

Reply:
Short term yes. Long term we could JOIN with the UTXO set table (once made readable). The table won't have the missing UTXO so they won't show up.

Reply:
Ok I'll add a `bool` field named `slashed`

Reply:
I think, we rely on the key_to_index and index_to_key maps in the Enrollment manager for block reward, because those maps are never cleared.

Reply:
> I think, we rely on the key_to_index and index_to_key maps in the Enrollment manager for block reward, because those maps are never cleared.

At heights after the validator is slashed we should not add it to those maps though. I think it will be safer to mark as slashed in the table and leave the records there for validating older block signatures in case of the edge cases where we still get one for a slashed validator from a previous block.

Reply:
> > I think, we rely on the key_to_index and index_to_key maps in the Enrollment manager for block reward, because those maps are never cleared.
> 
> At heights after the validator is slashed we should not add it to those maps though. 
I expected that when I was implementing block reward calculation, so it shouldn't cause problems.